crates/sargon/src/factor_instances_provider/provider/instances_in_cache_consumer.rs

```diff
@@ -0,0 +1,31 @@
+use crate::prelude::*;
+use futures::future::{BoxFuture, Future};
+
+/// A handle for consuming used instances that are present in the cache, recently read from the cache
+/// but not eagerly consumed since we depended on a subsequent operation to succeed before we would
+/// consume them.
+pub struct InstancesInCacheConsumer {
+    do_consume:
+        Box<dyn Fn() -> BoxFuture<'static, Result<()>> + Send + 'static>,
+}
+
+impl InstancesInCacheConsumer {
+    pub(super) fn new<T, F>(f: T) -> Self
+    where
+        T: Send + Sync + 'static + Fn() -> F,
+        F: Future<Output = Result<()>> + Send + 'static,
+    {
+        InstancesInCacheConsumer {
+            do_consume: Box::new(move || Box::pin(f())),
+        }
+    }
+
+    /// Consumes FactorInstances from the FactorInstancesCache, call this method
+    /// once you have successfully used the FactorInstance(s), e.g. saved a newly
+    /// created virtual entity into Profile, or successfully secured an entity,
+    /// that is submitted a transaction updating the on chain rules and also
+    /// successfully updated it in Profile and saved Profile to secure storage.
+    pub async fn consume(self) -> Result<()> {
```

Comment:
As you have identified, it is a non issue for FI in Unsecurified space => same entity address => adding said entity will fail. with `Err(CommonError::UnableToAddAllAccountsDuplicatesFound)`

We SHOULD we ensure to do the same for MFA, that is, when we call `update_entity` we should inspect ALL FIs of ALL entities and ensure the FI is not already used.

Reply:
I will start looking into harmonizing `create_account` and (new) `create_persona` and `sargon_os_accounts` and (new) `sargon_os_personas` to use generics, a single shared impl for all APIs, which make it easier for us to do this `ensure_factor_instance_is_new` check inside of (new, generic impl of `update_account`/`update_persona`) `update_entity`

Reply:
why it should be called also from `replace_profile_state_with` and `update_profile_with`?

Reply:
@giannis-rdx we have some normalization work to do... there are multiple execution paths which replaces ProfileStateHolder's state with a new Profile. I wanna "capture them all", which would be trivially done - in one place - if we had refactored the code a bit to always go through on single method. To Be Done later. So for now I just try to make sure we always assert that all factor instances are "valid" (unique is a slightly misleading word, because for Securified Entities the same FactorInstance can be used multiple times in different roles. )